basic_features/basic_mod_data_checker.py

```diff
@@ -0,0 +1,258 @@
+# -*- encoding: utf-8 -*-
+from __future__ import annotations
+
+import fnmatch
+import re
+import sys
+from collections.abc import Iterable, Sequence
+from typing import MutableMapping, Optional, TypedDict, overload
+
+import mobase
+
+
+def convert_entry_to_tree(entry: mobase.FileTreeEntry) -> Optional[mobase.IFileTree]:
+    if not entry.isDir():
+        return None
+    if isinstance(entry, mobase.IFileTree):
+        return entry
+    if (parent := entry.parent()) is None:
+        return None
+    converted_entry = parent.find(
+        entry.name(), mobase.FileTreeEntry.FileTypes.DIRECTORY
+    )
+    if isinstance(converted_entry, mobase.IFileTree):
+        return converted_entry
+    return None
+
+
+class RegexPatternDict(dict, MutableMapping[str, re.Pattern]):
+    """Regex patterns for validation in `BasicModDataChecker`."""
+
+    @classmethod
+    def from_glob_patterns(cls, glob_patterns: GlobPatternDict) -> RegexPatternDict:
+        """Returns an instance of `RegexPatternDict`, with the `glob_patterns`
+        translated to regex.
+        """
+        return cls(
+            (
+                key,
+                cls.regex_from_glob_list(value)
+                if isinstance(value, Iterable)
+                else None,
+            )
+            for key, value in glob_patterns.items()
+        )
+
+    def get_match_index(self, key: str, search_str: str) -> Optional[int]:
+        """Get the index of the matched group if the `self[key]` matches `search_str`.
+
+        Returns:
+            The 0-based index of the matched group or None for no match.
+        """
+        if (pattern := self.get(key)) is None:
+            return None
+        else:
+            return self.match_index_of_pattern(search_str, pattern)
+
+    @staticmethod
+    def match_index_of_pattern(search_str: str, pattern: re.Pattern) -> Optional[int]:
+        """Get the index of the matched group if `search_str` matches `pattern`
+        (from this dict).
+
+        Returns:
+            The 0-based index of the matched group or None for no match.
+        """
+        if not (match := pattern.match(search_str)) or match.lastindex is None:
```

Comment:
I don't really like these `xxx := foo` that use `xxx` outside of the block it was defined, this does not bring anything in term of readability or length and add confusion.

Reply:
Inverting it makes it clearer: (`match.lastindex` should never be 0...)
```py
if (match := pattern.match(search_str)) and match.lastindex:
    return match.lastindex - 1
return None
```
But if you insist:
```py
match = pattern.match(search_str)
if match and match.lastindex:
    return match.lastindex - 1
return None
```